libr/util/idpool.c

```diff
@@ -195,3 +200,263 @@ R_API void r_id_storage_free(RIDStorage* storage) {
 	}
 	free (storage);
 }
+
+R_API ROIDStorage * r_oids_new (ut32 start_id, ut32 last_id) {
+	ROIDStorage *storage = R_NEW0 (ROIDStorage);
+	if (!storage) {
+		return NULL;
+	}
+	if (!(storage->permutation = r_id_storage_new (0, last_id))) {
+		free (storage);
+		return NULL;
+	}
+	if (!(storage->data = r_id_storage_new (start_id, last_id))) {
+		r_id_storage_free (storage->permutation);
+		free (storage);
+		return NULL;
+	}
+	return storage;
+}
+
+R_API void *r_oids_get(ROIDStorage *storage, ut32 id) {
+	if (storage) {
+		return r_id_storage_get (storage->data, id);
+	}
+	return NULL;
+}
+
+R_API void *r_oids_oget(ROIDStorage *storage, ut32 od) {
+	ut32 id;
+	if (r_oids_get_id (storage, od, &id)) {
+		return r_id_storage_get (storage->data, id);
+	}
+	return r_id_storage_get (storage->data, id);
+}
+
+R_API bool r_oids_get_id(ROIDStorage *storage, ut32 od, ut32 *id) {
+	ut32 *_id;
+	if (storage && storage->permutation && (od < storage->permutation->pool->next_id) &&
+		(_id = storage->permutation->data[od])) {
+		*id = *_id;
+		return true;
+	}
+	return false;
+}
+
+R_API bool r_oids_get_od(ROIDStorage *storage, ut32 id, ut32 *od) {
+	ut32 *_id;
+	if (storage && storage->permutation &&
+		storage->data && (id < storage->data->pool->next_id)) {
+		for (*od = 0; *od < storage->permutation->pool->next_id; *od++) {
+			_id = (ut32 *)storage->permutation->data[*od];
+			if (_id && (id == *_id)) {
+				return true;
+			}
+		}
+	}
+	return false;
+}
+
+R_API bool r_oids_add(ROIDStorage *storage, void *data, ut32 *id, ut32 *od) {
+	ut32 *_id;
+	if (!storage || !id || !od || !(_id = R_NEW (ut32))) {
```

Comment:
R_NEW(ut32)? why not use a pool instead, thats unnefficient

Reply:
bc this way we can use _id = NULL for error somewhere else in the code. also void * != ut32